There are nine variables in the dataset as follows:

name: Name of the blend    
roaster: Name of the roaster    
roast: Type of roast (Light, Medium-Light, Medium, Medium-Dark, Dark)    
loc_country: Location of the roaster    
origin: Origin of the beans    
100g_USD: Price per 100g of beans in US dolalrs    
rating: Rating of the coffee    
review_date: Date of the coffee review    
review: Text of review    

Import libraries and dataset

In [5]:
import pandas as pd

df = pd.read_csv('simplified_coffee.csv')
df


,name,roaster,roast,loc_country,origin,100g_USD,rating,review_date,review
0,Ethiopia Shakiso Mormora,Revel Coffee,Medium-Light,United States,Ethiopia,4.70,92,November 2017,"Crisply sweet, cocoa-toned. Lemon blossom, roa..."
1,Ethiopia Suke Quto,Roast House,Medium-Light,United States,Ethiopia,4.19,92,November 2017,"Delicate, sweetly spice-toned. Pink peppercorn..."
2,Ethiopia Gedeb Halo Beriti,Big Creek Coffee Roasters,Medium,United States,Ethiopia,4.85,94,November 2017,"Deeply sweet, subtly pungent. Honey, pear, tan..."
3,Ethiopia Kayon Mountain,Red Rooster Coffee Roaster,Light,United States,Ethiopia,5.14,93,November 2017,"Delicate, richly and sweetly tart. Dried hibis..."
4,Ethiopia Gelgelu Natural Organic,Willoughby's Coffee & Tea,Medium-Light,United States,Ethiopia,3.97,93,November 2017,"High-toned, floral. Dried apricot, magnolia, a..."
...,...,...,...,...,...,...,...,...,...
1241,Finca Patzibir,El Gran Cafe,Medium-Light,Guatemala,Guatemala,4.70,92,November 2022,"Crisply sweet, nut-toned. Almond brittle, pie ..."
1242,Proyecto Aurora,El Gran Cafe,Medium-Light,Guatemala,Guatemala,2.94,93,November 2022,"Chocolaty, floral-framed. Cocoa nib, honeysuck..."
1243,Finca El Potrero,El Gran Cafe,Medium-Light,Guatemala,Guatemala,2.94,93,November 2022,"High-toned, enticingly sweet. Black cherry, na..."
1244,Chacayá Santiago Atitlán,El Gran Cafe,Medium-Light,Guatemala,Guatemala,4.70,93,November 2022,"Vibrantly sweet, subtly nuanced. Apricot, dar..."


Get an overview of the data in the next few cells.

In [2]:
df.head()

,name,roaster,roast,loc_country,origin,100g_USD,rating,review_date,review
0,Ethiopia Shakiso Mormora,Revel Coffee,Medium-Light,United States,Ethiopia,4.70,92,November 2017,"Crisply sweet, cocoa-toned. Lemon blossom, roa..."
1,Ethiopia Suke Quto,Roast House,Medium-Light,United States,Ethiopia,4.19,92,November 2017,"Delicate, sweetly spice-toned. Pink peppercorn..."
2,Ethiopia Gedeb Halo Beriti,Big Creek Coffee Roasters,Medium,United States,Ethiopia,4.85,94,November 2017,"Deeply sweet, subtly pungent. Honey, pear, tan..."
3,Ethiopia Kayon Mountain,Red Rooster Coffee Roaster,Light,United States,Ethiopia,5.14,93,November 2017,"Delicate, richly and sweetly tart. Dried hibis..."
4,Ethiopia Gelgelu Natural Organic,Willoughby's Coffee & Tea,Medium-Light,United States,Ethiopia,3.97,93,November 2017,"High-toned, floral. Dried apricot, magnolia, a..."


Explore the columns

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1246 entries, 0 to 1245
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1246 non-null   str    
 1   roaster      1246 non-null   str    
 2   roast        1234 non-null   str    
 3   loc_country  1246 non-null   str    
 4   origin       1246 non-null   str    
 5   100g_USD     1246 non-null   float64
 6   rating       1246 non-null   int64  
 7   review_date  1246 non-null   str    
 8   review       1246 non-null   str    
dtypes: float64(1), int64(1), str(7)
memory usage: 87.7 KB


In [3]:
df.shape

(1246, 9)

There might be 12 null values in the roast column.

Now I need to check data types, which I can see above, but I will write code for it anyway.

In [4]:
print(df.dtypes)

name               str
roaster            str
roast              str
loc_country        str
origin             str
100g_USD       float64
rating           int64
review_date        str
review             str
dtype: object


There is a good mix of strings and numbers (floats and integers).  I might need to separate the individual words out of the review column for analysis.    

Next, I will check for null values.  

In [5]:
df.isnull()

,name,roaster,roast,loc_country,origin,100g_USD,rating,review_date,review
0,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...
1241,False,False,False,False,False,False,False,False,False
1242,False,False,False,False,False,False,False,False,False
1243,False,False,False,False,False,False,False,False,False
1244,False,False,False,False,False,False,False,False,False


Since the df only shows me the first and last 5 records, I don't know which rows contain the null values.

In [8]:
df.isna()

,name,roaster,roast,loc_country,origin,100g_USD,rating,review_date,review
0,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...
1241,False,False,False,False,False,False,False,False,False
1242,False,False,False,False,False,False,False,False,False
1243,False,False,False,False,False,False,False,False,False
1244,False,False,False,False,False,False,False,False,False


df.isna() and df.isnull() do the same thing and don't tell me what I need to know, which is which rows contain the null values.   

Next, I will try this:

In [9]:
null_rows = df[df['roast'].isnull()]
print(null_rows)

                                                   name  \
132                      Esmeralda Estate Panama Geisha   
480                      Esmeralda Estate Panama Geisha   
481               Asfaw Maru Ethiopia Natural Cold Brew   
482                      Ethiopia Nano Challa Cold Brew   
483        Brazil Ipanema Black Edition A-41 Red Cherry   
606                      Esmeralda Estate Panama Geisha   
822            Brazil Ipanema Golden Edition C26 Lychee   
826                      Esmeralda Estate Panama Geisha   
882   Ethiopia Guji Hambela Wate 74110/74112/74158 M...   
886                                       Yemen Lot 106   
892       Ethiopia Guji Shakisso Mormora Estate Natural   
1131                     Esmeralda Estate Panama Geisha   

                     roaster roast    loc_country    origin  100g_USD  rating  \
132        Difference Coffee   NaN        England    Panama    111.11      94   
480        Difference Coffee   NaN        England    Panama    111.11 

There are twelve null rows.  I am going to drop them since it's just a few rows and I have enough data otherwise.  I could fill in the missing data with the most common value (type of roast) but it might be inaccurate.  To drop the rows, I need to write dropna() and assign it to a variable.       

According to W3 Schools, "The dropna() method returns a new DataFrame object unless the inplace parameter is set to True, in that case the dropna() method does the removing in the original DataFrame instead." I originally did inplace = True before I discovered it was best to assign it to a variable.  I might have modified the original DataFrame.       

In [12]:
df = df.dropna()

Determine if the dropna() was successful: 

In [13]:
df.shape

(1234, 9)

As you can see, the dropna() was successful, as there are now only 1234 rows.  

Now, I will check again for any null values in the entire DataFrame.

In [14]:
df.isnull().sum()

name           0
roaster        0
roast          0
loc_country    0
origin         0
100g_USD       0
rating         0
review_date    0
review         0
dtype: int64

There are no null values in the DataFrame.  

Now I need to check for unique values in the DataFrame. It will return a Series containing counts of unique values.

In [15]:
df['name'].value_counts()

name
Ethiopia Kayon Mountain                  6
Kona Mocca®                              5
100% Kona Bourbon Pointu Laurina         4
Bella Carmona Guatemala                  4
Guatemala Bella Carmona                  4
                                        ..
Ethiopia Guji Hambela Wate Natural G1    1
Finca Patzibir                           1
Finca El Potrero                         1
Chacayá Santiago Atitlán                 1
Espresso No. 3                           1
Name: count, Length: 1134, dtype: int64

In [16]:
df['roaster'].value_counts()

roaster
Kakalove Cafe                       91
JBC Coffee Roasters                 91
Paradise Roasters                   83
Dragonfly Coffee Roasters           30
Bird Rock Coffee Roasters           29
                                    ..
JYL Cafe                             1
Balmy Day Coffee Office              1
Dr. Young Cafe                       1
San Francisco Bay Coffee Company     1
Durango Coffee Companuy              1
Name: count, Length: 298, dtype: int64

In [17]:
df['roast'].value_counts()

roast
Medium-Light    904
Light           198
Medium          114
Medium-Dark      15
Dark              3
Name: count, dtype: int64

In [18]:
df['loc_country'].value_counts()

loc_country
United States    771
Taiwan           334
Hawai'i           77
Guatemala         24
Hong Kong          9
Japan              8
Canada             5
England            2
Australia          1
China              1
Kenya              1
New Taiwan         1
Name: count, dtype: int64

In [19]:
df['origin'].value_counts()

origin
Ethiopia                            443
Colombia                            151
Kenya                               143
Hawai'I                             113
Guatemala                            96
Panama                               65
Costa Rica                           46
Indonesia                            28
Democratic Republic Of The Congo     14
El Salvador                          14
Peru                                 13
Brazil                               13
Rwanda                               12
Burundi                              12
Yemen                                12
Nicaragua                            11
Mexico                               11
Taiwan                                9
Ecuador                               8
Honduras                              6
Nepal                                 2
Bolivia                               2
Dominican Republic                    2
Thailand                              2
Uganda                           

In [ ]:
df['100g_USD'].value_counts()

100g_USD
5.29     68
5.58     51
5.88     50
6.47     30
6.17     29
         ..
9.04      1
1.36      1
14.16     1
6.08      1
8.11      1
Name: count, Length: 338, dtype: int64

In [21]:
df['rating'].value_counts()

rating
93    356
94    349
92    184
95    177
91     60
90     43
96     43
97     10
89      3
88      3
87      3
86      2
84      1
Name: count, dtype: int64

In [22]:
df['review_date'].value_counts()

review_date
October 2020     51
November 2021    38
November 2020    36
June 2022        35
October 2022     35
                 ..
March 2020        8
June 2018         7
January 2020      7
January 2019      6
July 2018         3
Name: count, Length: 61, dtype: int64

In [23]:
df['review'].value_counts()

review
Crisply sweet, cocoa-toned. Lemon blossom, roasted cacao nib, date, rice candy, white peppercorn in aroma and cup. Savory-tart structure; delicate, silky mouthfeel. The richly drying finish leads with cocoa-toned, crisply sweet floral notes in the short and a hint of spice (white peppercorn) in the long.                     1
Delicate, sweetly spice-toned. Pink peppercorn, date, myrrh, lavender, roasted cacao nib in aroma and cup. Crisp, spice-toned structure with citrus-like acidity; satiny, very smooth mouthfeel. The crisply sweet finish centers around spice and pungent floral notes.                                                              1
Deeply sweet, subtly pungent. Honey, pear, tangerine zest, dark chocolate, pistachio in aroma and cup. Sweet and juicy yet crisp in structure (think ripe pear); plush and buoyant in mouthfeel. Resonant and flavor saturated in the short finish; simplifies around hints of tangerine and pistachio in the long.                   1
Delicate,

Now, I need to get an overall description of the data in the DataFrame.

In [24]:
df.describe()

,100g_USD,rating
count,1234.000000,1234.000000
mean,10.093679,93.300648
std,11.965835,1.483475
min,0.170000,84.000000
25%,5.280000,93.000000
50%,6.170000,93.000000
75%,9.600000,94.000000
max,132.280000,97.000000
